toml/src/main/java/com/fasterxml/jackson/dataformat/toml/Parser.java

```
@@ -0,0 +1,407 @@
+                return factory.textNode(text);
+            case TRUE:
+                pollExpected(TomlToken.TRUE, nextState);
+                return factory.booleanNode(true);
+            case FALSE:
+                pollExpected(TomlToken.FALSE, nextState);
+                return factory.booleanNode(false);
+            case OFFSET_DATE_TIME:
+            case LOCAL_DATE_TIME:
+            case LOCAL_DATE:
+            case LOCAL_TIME:
+                return parseDateTime(nextState);
+            case FLOAT:
+                return parseFloat(nextState);
+            case INTEGER:
+                return parseInt(nextState);
+            case ARRAY_OPEN:
+                return parseArray(nextState);
+            case INLINE_TABLE_OPEN:
+                return parseInlineTable(nextState);
+            default:
+                throw errorContext.atPosition(lexer).unexpectedToken(firstToken, "value");
+        }
+    }
+
+    private JsonNode parseDateTime(int nextState) throws IOException {
+        String text = lexer.yytext();
+        TomlToken token = poll(nextState);
+        // the time-delim index can be [Tt ]. java.time supports only [Tt]
+        if ((token == TomlToken.LOCAL_DATE_TIME || token == TomlToken.OFFSET_DATE_TIME) && text.charAt(10) == ' ') {
+            text = text.substring(0, 10) + 'T' + text.substring(11);
+        }
+
+        if (options.parseTemporalAsJavaTime) {
+            Temporal value;
+            if (token == TomlToken.LOCAL_DATE) {
+                value = LocalDate.parse(text);
+            } else if (token == TomlToken.LOCAL_TIME) {
+                value = LocalTime.parse(text);
+            } else {
+                if (token == TomlToken.LOCAL_DATE_TIME) {
+                    value = LocalDateTime.parse(text);
+                } else if (token == TomlToken.OFFSET_DATE_TIME) {
+                    value = OffsetDateTime.parse(text);
+                } else {
+                    throw new AssertionError();
```

I'd prefer converting all assert/AssertionErrors into different kind of internal exception, just something that is not `Error` (since I think catching `Error` is usually no-no, compared to `Exception`).
Jackson-core has `VersionUtil.throwInternal()` which throws a RuntimeException, for example, but need not use that one. Probably best to first add a helper method in this class, call that, let it throw whatever is deemed best.

---

the code path should never be executed, so throwInternal sounds good.